### PR TITLE
feat(restPassword):forgot password

### PR DESCRIPTION
**Pull request by IRASUBIZA-Elyse**
#### What does this PR do?

resetting password via email

#### Description of Task to be completed?

- send a forgot password request
- reset the password via eamil using the link sent to the email

#### How should this be manually tested?

- valid the email used to reset password
- use the link sent to the email and reset the password

#### What are the relevant pivotal tracker/Trello stories?

- id of the project
 #187300149

### CODE REVIEW

**Review by Nadinefiona — APPROVED** [review]

File: src/controllers/resetPasswordController.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import { Request, Response } from 'express';
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+import bcrypt from 'bcrypt';
```
Comment (fab-ryan) [inline src/controllers/resetPasswordController.ts]: we are using brcyptjs not this

File: src/controllers/resetPasswordController.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        'A user with this email does not exist.'
+      );
+    }
+    const resetToken = jwt.sign(
+      { id: user.id },
+      process.env.JWT_SECRET || 'SECRET',
+      {
+        expiresIn: '1h',
+      }
+    );
```
Comment (fab-ryan) [inline src/controllers/resetPasswordController.ts]: we already have token generator function

File: src/controllers/resetPasswordController.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+export const resetPassword = async (req: Request, res: Response) => {
+  const { token } = req.params;
+  const { password } = req.body;
+
+  try {
+    const decoded = jwt.verify(token, process.env.JWT_SECRET || 'secret') as {
+      id: string;
+      exp: number;
+    };
```
Comment (fab-ryan) [inline src/controllers/resetPasswordController.ts]: and also this

File: src/routers/userRoute.ts [hunk 4]
```diff
@@ -25,5 +33,10 @@ router.post(
   validationMiddleware(changePasswordValidationSchema),
   updatePassword
 );
-
+router.post('/users/forgot-password', forgotPassword);
```
Comment (fab-ryan) [inline src/routers/userRoute.ts]: validation of email is required

File: src/utils/sendEmails.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  const mailOptions = {
+    from: process.env.EMAIL_USER,
+    to: email,
+    subject: 'Resetting password',
+    html: `<p>Click on the link to reset your password: <a href="${link}">Reset Password</a></p>`,
+  };
+
+  await transporter.sendMail(mailOptions);
+}
+export default sendResetPasswordEmail;
```
Comment (fab-ryan) [inline src/utils/sendEmails.ts]: this whole function of sending email is build already please first ready the code before starting implementations

File: src/validations/userValidation.ts [hunk 6]
```diff
@@ -31,3 +31,17 @@ export const loginUserSchema = Joi.object({
...
 });
+export const resetPasswordSchema = Joi.object({
+  password: Joi.string()
+    .required()
+    .custom((value, helpers) => {
+      const passwordError = testPassword(value);
+      if (passwordError) {
+        return helpers.error('any.invalid', { message: passwordError });
+      }
+      return value;
```
Comment (fab-ryan) [inline src/validations/userValidation.ts]: avoid using this method please use regix methods

**Review by fab-ryan — CHANGES_REQUESTED** [review]
I have remained with comments

**Review by musabehonore — APPROVED** [review]

**Review by Mukunzijames — APPROVED** [review]

File: src/controllers/resetPasswordController.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    await user.update({
+      password: hashedPassword,
+    });
+    const data = signToken({ id: user.id }, '2h');
+    res.cookie('token', data);
+    res.header('Authorization', `Bearer ${data}`);
+    return sendResponse<string>(
+      res,
+      200,
+      data,
```
Comment (fab-ryan) [inline src/controllers/resetPasswordController.ts:87]: you shouldn't have set data to cookies
and also should be token not data

**Review by fab-ryan — CHANGES_REQUESTED** [review]
Please change that lines

**Review by fab-ryan — APPROVED** [review]

**Pull request merged**